File: 1884-minimum-changes-to-make-alternating-binary-string/minimum-changes-to-make-alternating-binary-string.py
class Solution:
    def minOperations(self, s: str) -> int:
        # start with 0, odd == 0, even == 1
        # start with 1, odd == 1, even == 0
        ops_w_zero_start, ops_w_one_start = 0, 0
        for i in range(len(s)):
            if i % 2 == 0:  # even idx
                if s[i] == '0':
                    ops_w_zero_start += 1
                else:
                    ops_w_one_start += 1
            else:       # odd idx
                if s[i] == '0':
                    ops_w_one_start += 1
                else:
                    ops_w_zero_start += 1
        return min(ops_w_zero_start, ops_w_one_start)

File: 1884-minimum-changes-to-make-alternating-binary-string/minimum-changes-to-make-alternating-binary-string.java
class Solution {
    public int minOperations(String s) {
        // zeros at odd, ones at even position
        // score = total - zeros_odd - ones_even
        // score = total - ones_odd - zeros_even

        int zeros_odd = 0, zeros_even = 0, ones_odd = 0, ones_even = 0;
        for (int i = 0; i < s.length(); i++) {
            int digit = s.charAt(i) - '0';
            if ((i & 1) == 1) {
                zeros_odd += 1 ^ digit;
                ones_odd += digit;
            } else {
                zeros_even += 1 ^ digit;
                ones_even += digit;
            }
        }

        return s.length() - Math.max(zeros_odd + ones_even, zeros_even + ones_odd);
    }
}